src/Streamly/External/ByteString.hs

```diff
@@ -14,38 +14,67 @@ where
 
 import Control.Monad.IO.Class (MonadIO)
 import Data.Word (Word8)
-import GHC.ForeignPtr (ForeignPtr(..))
-import GHC.Ptr (Ptr(..), minusPtr, nullPtr, plusPtr)
-import Streamly.Data.Unfold (Unfold, lmap)
+import Foreign.ForeignPtr (withForeignPtr)
+import Foreign.Storable (peek)
+import GHC.Exts
+    ( Addr#
+    , MutableByteArray#
+    , RealWorld
+    , byteArrayContents#
+    , minusAddr#
+    , plusAddr#
+    , unsafeCoerce#
+    )
+import GHC.ForeignPtr (ForeignPtr(..), ForeignPtrContents(..))
+import GHC.Int (Int(..))
+import GHC.Ptr (Ptr(..), nullPtr, plusPtr)
 import Streamly.Data.Fold (Fold)
+import Streamly.Data.Unfold (Unfold, lmap)
 
 -- Internal imports
 import Data.ByteString.Internal (ByteString(..))
-import Streamly.Internal.Data.Array.Foreign.Type (Array(..))
-import Streamly.Internal.Data.Array.Foreign.Mut.Type
-    (ArrayContents, arrayToFptrContents, fptrToArrayContents, nilArrayContents)
+import Streamly.Internal.Data.Array.Type (Array(..))
+import Streamly.Internal.Data.Unboxed (MutableByteArray(..))
+import Streamly.Internal.System.IO (unsafeInlineIO)
 
-import qualified Streamly.Data.Array.Foreign as A
+import qualified Streamly.Data.Array as Array
+import qualified Streamly.Internal.Data.Unboxed as Unboxed
+import qualified Streamly.Internal.Data.Unfold as Unfold
+import qualified Streamly.Internal.Data.Stream.StreamD as StreamD
 
 import Prelude hiding (read)
 
+{-# INLINE mutableByteArrayContents# #-}
+mutableByteArrayContents# :: MutableByteArray# RealWorld -> Addr#
+mutableByteArrayContents# marr# = byteArrayContents# (unsafeCoerce# marr#)
+
 -- | Helper function that creates a ForeignPtr
-makeForeignPtr :: ArrayContents -> Ptr a -> ForeignPtr a
-makeForeignPtr contents (Ptr addr#) =
-    ForeignPtr addr# (arrayToFptrContents contents)
+{-# INLINE makeForeignPtr #-}
+makeForeignPtr :: MutableByteArray -> Int -> ForeignPtr a
+makeForeignPtr (MutableByteArray marr#) (I# off#) =
+    ForeignPtr
+        (mutableByteArrayContents# marr# `plusAddr#` off#)
+        (PlainPtr marr#)
 
 -- | Convert a 'ByteString' to an array of 'Word8'. This function unwraps the
 -- 'ByteString' and wraps it with 'Array' constructors and hence the operation
 -- is performed in constant time.
```

Comment:
It can be done in constant time only for GHC allocated memory. For foreign allocator allocated memory there is a copy involved.